Add `Form` docs (#611)
radix-ui/website: data/primitives/components/form/0.0.1.mdx

    },
  ]}
/>

## Examples

### Composing with your own components

Using `asChild` you can compose the `Form` primitive parts with your own components.

```jsx line=3
<Form.Field name="name">
  <Form.Label>Full name</Form.Label>
  <Form.Control __asChild__>
    <TextField variant="primary" />
  </Form.Control>
</Form.Field>
```

It can also be used to compose other types of controls, such as a `select`:

```jsx line=3
<Form.Field name="country">
  <Form.Label>Country</Form.Label>
  <Form.Control __asChild__>
    <select>
      <option value="uk">United Kingdom</option>…
    </select>
  </Form.Control>
</Form.Field>
```

### Providing your own validation messages

When no `children` are provided, `Form.Message` will render a default error message for the given `match`.

```jsx
// will yield "This value is missing"
<Form.Message match="missingValue" />
```

You can provide a more meaningful message by providing your own `children`. This is also useful for internationalization.

```jsx
// will yield "Please provide a name"
<Form.Message match="missingValue">__Please provide a name__</Form.Message>
```

### Custom validation

> On top of all the built-in client-side validation matches described above you can also provide your own custom validation whilst still making use of the platform's validation abilities. It uses the `customError` type present in the constraint validity API.

Reviewer: "constraint validity API" can we link to it?

> You can pass your own validation function into the `match` prop on `Form.Message`. Here's a contrived example:

Reviewer: ```suggestion
You can pass your own validation function into the `match` prop on `Form.Message`. Here's an example:
```
nit: imo contrived has a negative connotation without adding much.

```jsx line=4
<Form.Field name="name">
  <Form.Label>Full name</Form.Label>
  <Form.Control />
  <Form.Message __match__={(value, formData) => value !== 'John'}>
    Only John is allowed.
  </Form.Message>
</Form.Field>
```

> `match` will be called with the current value of the control as first argument and the entire `FormData` as second argument.

> > `match` can also be an `async` function (or return a promise) to perform async validation (see type below):

Reviewer: "see type below" – what is this referring to?

Author: old copy/paste

### Styling based on validity

We add `data-valid` and `data-invalid` attributes to the relevant parts. Use it to style your components accordingly.
Here is an example styling the `Label` part.

```jsx line=8
//index.jsx
import * as React from 'react';
import Form from '@radix-ui/react-form';

export default () => (
  <Form.Root>
    <Form.Field name="email">
      <Form.Label __className__="FormLabel">Email</Form.Label>
      <Form.Control type="email" />
    </Form.Field>
  </Form.Root>
);
```

```css line=2
/* styles.css */

> .FormLabel[__data-invalid__] {

Reviewer: For completeness can we add `data-valid` with `color: green`?

  color: red;
}
```

### Accessing the validity state for more control

> Sometimes, you may need to access the raw validity state of a field in order to display your own icons, or interface with your own component library by passing certain props to it. You can do this by using the `Form.ValidityState` part:

Reviewer: ```suggestion
You may need to access the raw validity state of a field in order to display your own icons, or interface with a component library via it's defined props. You can do this by using the `Form.ValidityState` part:
```

```jsx line=3-4,7-8
<Form.Field name="name">
  <Form.Label>Full name</Form.Label>
  <Form.ValidityState>
    {(validity) => (
      <Form.Control asChild>

>         <TextField variant="primary" state={getTextFieldState(__validity__)} />

Reviewer: ```suggestion
        <TextField variant="primary" error={__validity__.valueMissing} />
```

I feel like this line is too opaque and we should be clearer on what's happening in the implementation. My suggestion is not the best but do you think it would help to be more explicit on what `getTextFieldState` is doing?

Author: I'm not sure it would to be honest, and I'm not sure you need to know what `getTextFieldState` does, the important thing is that it's `state = fn(validity)` so `validity` is needed as a prop somehow.

Reviewer: Yeah you're probably right, I'm might be over thinking what we need to be demonstrating.

      </Form.Control>
    )}
  </Form.ValidityState>
</Form.Field>
```

### Server-side validation

The component also supports server-side validation using the same `Form.Message` component.

> You can re-use the same messages you defined for client-side errors by passing a `forceMatch` prop which will for the message to show regardless of the client-side matching logic.

Reviewer: ```suggestion
You can re-use the same messages you defined for client-side errors by passing a `forceMatch` prop which will force the message to show regardless of the client-side matching logic.
```

If the message doesn't exist on the client-side, you can render a `Form.Message` without a `match` too.
The field is marked as invalid by passing a `serverInvalid` boolean prop to the `Form.Field` part.

Here's an example with server-side error handling:

```jsx line=12,23,29-34
import * as React from 'react';
import * as Form from '@radix-ui/react-form';

function Page() {
  const [serverErrors, setServerErrors] = React.useState({
    email: false,
    password: false,
  });

  return (
    <Form.Root
      onSubmit={(event) => {
        // `onSubmit` only triggered if it passes client-side validation

>         event.preventDefault();

Reviewer: I find these two lines confusing, is the comment related to the `preventDefault`? it reads like it is yet the comment doesn't match my understanding of the `preventDefault` usage here.

Author: Yeah I can probably remove the comment.

Author: I've reworked it differently.

        const data = Object.fromEntries(new FormData(event.currentTarget));

>         // maybe do an async server call and gather some errors from the server to display

Reviewer: ```suggestion
        // Submit form data and catch errors in the response
```

        setServerErrors({ email: false, password: false })
      }
    >
      <Form.Field name="email" __serverInvalid__={serverErrors.email}>
        <Form.Label>Email address</Form.Label>
        <Form.Control type="email" required />
        <Form.Message match="missingValue">
          Please enter your email.
        </Form.Message>
        <Form.Message match="typeMismatch" __forceMatch__={serverErrors.email}>
          Please provide a valid email.
        </Form.Message>
      </Form.Field>

      <Form.Field name="password" __serverInvalid__={serverErrors.password}>
        <Form.Label>Password</Form.Label>
        <Form.Control type="password" required />
        <Form.Message match="missingValue">
          Please enter a password.
        </Form.Message>
        {serverErrors.password && (
          <Form.Message>
            Please provide a valid password. It should contain at least 1 number
            and 1 special character.
          </Form.Message>
        )}
      </Form.Field>

      <Form.Submit>Submit</Form.Submit>
    </Form.Root>
  );
}
```

You should clear the server errors using the `onClearServerErrors` callback prop on the `Form.Root` part.

> > This will clear the server errors before the form is re-submitted, and when the form is reset.

Reviewer: Would this read better as part of the original paragraph?

> On top of this, it means you have control over when to reset single server errors.

Reviewer: ```suggestion
In addition, this provides control over when to reset single server errors.
```

For example you could reset the email server error as soon as the user edits it:

```jsx line=3
<Form.Field name="email" serverInvalid={serverErrors.email}>
  <Form.Label>Email address</Form.Label>
  <Form.Control
    type="email"
    __onChange__={() => setServerErrors((prev) => ({ ...prev, email: false }))}
  />
  <Form.Message match="missingValue">Please enter your email.</Form.Message>
  <Form.Message match="typeMismatch" forceMatch={serverErrors.email}>
    Please provide a valid email.
  </Form.Message>
</Form.Field>
```

## Accessibility

The component follows the "inline errors" pattern for validation:

- Label and control are associated using the `name` provided on `Form.Field`

> - When one or more client-side error message display, they are automatically associated to their matching control and announced accordingly

Reviewer: ```suggestion
- When one or more client-side error messages display, they are automatically associated with their matching control and announced accordingly
```
